mlflow/utils/async_logging/async_logging_queue.py

```diff
@@ -0,0 +1,192 @@
+        """
+        self._lock.acquire()
+        if not self._active_run_data_batches.get(run_id, None):
+            self._active_run_data_batches[run_id] = Queue()
+        self._lock.release()
+        return self._active_run_data_batches[run_id]
+
+    def _at_exit_callback(self) -> None:
+        """
+        Callback function to be executed when the program is exiting.
+        Stops the data processing thread and waits for the
+        queue to be drained.
+        Finally, shuts down the thread pools used for data logging and batch processing status
+          check.
+        """
+        try:
+            # Stop the data processing thread
+            self.continue_to_process_data = False
+            # Waits till queue is drained.
+            self.run_data_process_thread.result()
+            _RUN_DATA_LOGGING_THREADPOOL.shutdown(wait=False)
+            _RUN_BATCH_PROCESSING_STATUS_CHECK_THREADPOOL.shutdown(wait=False)
+        except Exception as e:
+            _logger.error(f"Error while callback from atexit in _at_exit_callback: {e}")
+
+    def _logging_loop(self) -> None:
+        """
+        Continuously processes run data from the logging queue until `continue_to_process_data`
+         is False.
+        If an exception is raised during processing, it is logged and re-raised.
+        """
+        try:
+            while self.continue_to_process_data:
+                self._log_run_data()
+        except Exception as e:
+            from mlflow.exceptions import MlflowException
+
+            raise MlflowException(f"Exception inside the run data logging thread: {e}")
+
+    def _log_run_data(self) -> None:
+        """
+        Process the run data in the running runs queues.
+
+        For each run in the running runs queues, this method retrieves the next batch of run data
+         from the queue and processes it by calling the `_processing_func` method with the run ID,
+           metrics, parameters, and tags in the batch.
+        If the batch is empty, it is skipped. After processing the batch, the processed watermark
+         is updated and the batch event is set.
+        If an exception occurs during processing, the exception is logged and the batch event is set
+        with the exception. If the queue is empty, it is ignored.
+
+        Returns: None
+        """
+        run_batch = None  # type: RunBatch
+        for run_id, run_queue in self._active_run_data_batches.items():
+            while not run_queue.empty():
+                try:
+                    run_batch = run_queue.get(timeout=1)
+                except Empty:
+                    # Ignore empty queue exception
+                    continue
+                try:
+                    self._logging_func(
+                        run_id=run_id,
+                        metrics=run_batch.metrics,
+                        params=run_batch.params,
+                        tags=run_batch.tags,
```

Comment:
LogBatch only supports 100 tags, 100 params, and 1000 metrics per request. Do we handle the case where this call site passes more tags / params / metrics than that by splitting into multiple requests?

Reply:
This queue has no notion of limits on the processing the batch size. Hence the sender would preprocess anything exceeds the max limit here - https://github.com/sagarsumant/mlflow/blob/sasum/mlflow-async/mlflow/tracking/_tracking_service/client.py#L408

I can definitely add the test case to cover this batch size exceeded thing, but wanted to clarify current status. Let me know what you think.